### PR TITLE
torch.nn.utils.clip_grad_norm_: remove device syncs

### LINKED ISSUE #60691: torch.nn.utils.clip_grad_norm_: bad GPU utilization due to GPU-data-dependent control-flow
## Issue description

During CUDA training, using [`torch.nn.utils.clip_grad_norm_`](https://github.com/pytorch/pytorch/blob/d69c22dd61a2f006dcfe1e3ea8468a3ecaf931aa/torch/nn/utils/clip_grad.py#L9) negatively affects my GPU's utilization.

Specifically, the implementation contains 3 instances where the CPU control-flow depends on the result of GPU computations, forcing stream synchronizations:
1. `if total_norm.isnan() or total_norm.isinf():` ([link](https://github.com/pytorch/pytorch/blob/d69c22dd61a2f006dcfe1e3ea8468a3ecaf931aa/torch/nn/utils/clip_grad.py#L43)):
    - The result of `isnan` is copied back to host pageable memory, to apply standard boolean short-circuit logic.
    - If no short-circuit occurs (the standard finite case), the same is done for the result of `isinf`.
3. `if clip_coef < 1:` ([link](https://github.com/pytorch/pytorch/blob/d69c22dd61a2f006dcfe1e3ea8468a3ecaf931aa/torch/nn/utils/clip_grad.py#L57)):
    - The comparison is run on the device as it should be.
    - The final result is copied back to host pageable memory to pick a branch.

Since gradient clipping is typically performed before calling `Optimizer.step()`, the result is that the optimizer's GPU computations (or whatever is run after clipping) suffer from poor pipelining and thus poor GPU utilization.

## Code example

The following is a dummy but complete model training example using gradient norm clipping:
```python
import torch


def main():
    cpu = torch.device("cpu")
    cuda = torch.device("cuda:0")

    tensorboard_profile_dir = "/tmp/tensorboard_profile/"
    with torch.profiler.profile(
        activities=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA],
        schedule=torch.profiler.schedule(wait=0, warmup=2, active=1, repeat=1),
        with_stack=True,
        on_trace_ready=torch.profiler.tensorboard_trace_handler(tensorboard_profile_dir),
    ) as profile:

        batch_size = 1
        num_feat = 1500  # Increase this to make the cuda sync in clip_grad_norm_ more dramatic.
        model = torch.nn.Linear(
            in_features=num_feat, out_features=num_feat, bias=False, device=cuda
        )

        optimizer = torch.optim.SGD(model.parameters(), lr=1e-3)

        with torch.profiler.record_function("init data"):
            # Random input data, and fixed regression target.
            x = torch.randn(batch_size, num_feat, device=cuda)
            target = torch.full((num_feat,), 100.0, device=cuda)

        for _ in range(3):
            with torch.profiler.record_function("forward + loss"):
                y = model(x)
                l1_loss = (y - target).abs().mean()

            with torch.profiler.record_function("backward"):
                optimizer.zero_grad()
                l1_loss.backward()

            with torch.profiler.record_function("clip grad"):
                torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=0.1, norm_type="inf")

            optimizer.step()

            profile.step()

    print("Saved profile to", tensorboard_profile_dir)


if __name__ == "__main__":
    main()
```

Corresponding profile in tensorboard:
clip_grad_norm_sync_issue annotated

## Workaround

In my training code, to work around the issue I changed the implementation of `clip_grad_norm_`:
1. I removed the `isnan()` and `isinf()` checks (and crossed my toes that my gradient stays well-behaved).
2. I replaced the `if clip_coef < 1:` by an unconditional scaling:
  ```diff
  -if clip_coef < 1:
  -    [... multiply gradients by clip_coef ...]
  +clip_coef = torch.clamp(clip_coef, max=1.0)
  +[... multiply gradients by clip_coef ...]
  ```

Doing so removed all 3 control-flow dependencies, and gave me a nice boost to GPU utilization (85% -> 95%) and training speed.
Obviously doing 1. is a pretty ill-advised; maybe there's some mechanism to defer raising the exception / printing the warning till later, when the result of the computation is available?
2. on the other hand merely adds no-op computations in the case where no clipping is necessary (but makes performance more consistent I suppose!).

<details>

<summary>Tensorboard trace with the workarounds</summary>

image

</details>

## System Info

- PyTorch or Caffe: PyTorch
- How you installed PyTorch (conda, pip, source): docker, using the `pytorch/pytorch:1.9.0-cuda11.1-cudnn8-devel` image.
- OS: Linux, Ubuntu 20.04
- PyTorch version: 1.9.0
- Python version: 3.7.10
- CUDA/cuDNN version: NVIDIA Driver 460.80, CUDA 11.2, cuDNN 8
- GPU models and configuration: GeForce MX150

cc @ngimel @VitalyFedyunin @heitorschueroff

### PR DESCRIPTION
**Pull request by mautier**
Fixes #60691

### Changes

Per the discussion in the above issue, this PR makes 2 changes:
1. When `error_if_nonfinite=False`, the NaN/Inf checks are truly skipped, and no device synchronization occurs.
    - Additionally, when performing the checks, the 2 results are combined with `torch.logical_or` to incur only a single sync (instead of 2 in the happy/finite path).
2. The `clip_coef` conditional is removed, in favor of a call to `clamp(..., max=1.0)` and an unconditional multiplication.

### Testing

- The existing unit tests for `clip_grad_norm_` pass.
- I have manually profiled the example program from #60691, and verified that:
    - No synchronizations occur when using `error_if_nonfinite=False`.
    - A single synchronization occurs when using `error_if_nonfinite=True`.

### CODE REVIEW

**Comment by mautier** [comment]
I briefly looked into the `pr/pytorch-linux-bionic-rocm4.2-py3.6` CI failure: `test_reduce_stress_cuda` in `test_c10d_gloo.py` is the failing test; as far as I can tell it is not related to `clip_grad_norm_` in any way, so let me know how to proceed!

Secondly, I'm not sure if the testing is did is sufficient: is there a way to assert in a test than 0 or 1 device sync is occuring during the call to `clip_grad_norm_`? Or is my manual check sufficient?

**Comment by jbschlosser** [comment]
> I briefly looked into the `pr/pytorch-linux-bionic-rocm4.2-py3.6` CI failure: `test_reduce_stress_cuda` in `test_c10d_gloo.py` is the failing test; as far as I can tell it is not related to `clip_grad_norm_` in any way, so let me know how to proceed!

There are pre-existing failures for that test- we can ignore it for the purposes of this PR.

> Secondly, I'm not sure if the testing I did is sufficient: is there a way to assert in a test than 0 or 1 device sync is occuring during the call to `clip_grad_norm_`? Or is my manual check sufficient?

No way to assert on device syncs that I'm aware of - cc @ngimel for confirmation

**Review by jbschlosser — COMMENTED** [review]
Changes on the python side lgtm - thanks for the update! Two things:
* Unfortunately, there's currently a C++ reimplementation of this function in `torch/csrc/api/include/torch/nn/utils/clip_grad.h` that should be updated as well
* I'm a bit curious about the performance implications of the unconditional multiplication on CPU

@JackCaoG Does it make sense to apply similar updates on the XLA side?

**Comment by JackCaoG** [comment]
Thanks @jbschlosser , will take a look tmr.

**Comment by JackCaoG** [comment]
@jbschlosser I think pt/xla can also benefit from this change, so I will update our patch as well. We are thinking about removing the patch for `clip_grad_norm_` for a while since `item` call is removed but we didn't benchmark the result yet..

**Comment by JackCaoG** [comment]
Ah actually pt/xla already do the manual [`clapping`](https://github.com/pytorch/xla/blob/a010f8913c2eb86f664392a13e409a2ea11721c0/torch_xla/_patched_functions.py#L46) and check for [error_if_nonfinite](https://github.com/pytorch/xla/blob/a010f8913c2eb86f664392a13e409a2ea11721c0/torch_xla/_patched_functions.py#L40) already. Pt/xla should be fine.

**Comment by jbschlosser** [comment]
@mautier Would you be willing to update the C++ version in `torch/csrc/api/include/torch/nn/utils/clip_grad.h` as well?

**Comment by mautier** [comment]
@jbschlosser  Sorry for the radio silence; life things and a fried power supply have gotten in the way :upside_down_face: 

I have a patch for the C++ version that I'll push soon; unfortunately it appears that the C++ version has a slightly different API, in that it returns a `double` (instead of a scalar tensor like in python). So while I was able to eliminate the unwanted syncs and align the `error_if_nonfinite` semantics, there remains a final sync that cannot be removed without changing the function signature.

**Comment by jbschlosser** [comment]
> @jbschlosser Sorry for the radio silence; life things and a fried power supply have gotten in the way 🙃
> 
> I have a patch for the C++ version that I'll push soon; unfortunately it appears that the C++ version has a slightly different API, in that it returns a `double` (instead of a scalar tensor like in python). So while I was able to eliminate the unwanted syncs and align the `error_if_nonfinite` semantics, there remains a final sync that cannot be removed without changing the function signature.

No worries! RIP your power supply :/

Thanks for checking out the C++ version as well! It's unfortunate that the API differences make that last sync unremovable, but it does makes sense. We want to keep the current API, but a comment about the discrepancy resulting in a final unremovable sync wouldn't hurt.

**Comment by albanD** [comment]
cc @ngimel

File: torch/csrc/api/include/torch/nn/utils/clip_grad.h [hunk 1]
```diff
@@ -23,43 +29,55 @@ inline double clip_grad_norm_(
...
-      auto param_max = param.grad().data().abs().max().item().toDouble();
-      if (param_max > total_norm || std::isnan(param_max)) {
-        total_norm = param_max;
-      }
+      norms.emplace_back(param.grad().data().abs().max());
     }
+    total_norm_tensor = (norms.size() == 1) ? norms[0] : torch::max(torch::stack(norms));
   } else if (norm_type == 0) {
-    total_norm = static_cast<double>(params_with_grad.size());
+    total_norm_tensor = torch::full({}, static_cast<double>(params_with_grad.size()));
```
Comment (mautier) [inline torch/csrc/api/include/torch/nn/utils/clip_grad.h:47]: I'm not all that familiar with the C++ tensor creation APIs, is there a better way of creating a scalar tensor?

And on an unrelated note: order 0 norm is defined in [`torch.linalg.norm`](https://pytorch.org/docs/stable/generated/torch.linalg.norm.html#torch.linalg.norm) as the number of non-zero entries, but in this implementation it's just the number of parameters?

Comment (jbschlosser) [reply torch/csrc/api/include/torch/nn/utils/clip_grad.h:47]: `torch::full()` is fine here imo.

Yeah, it does look like there's a discrepancy between here and the order 0 norm definition in `torch.linalg.norm`. I guess there's an implicit assumption that all grads are nonzero here? Regardless, we should leave as-is for backwards compatibility.

**Review by jbschlosser — APPROVED** [review]
Thanks for the awesome fix & commenting! LGTM

Mind rebasing to address merge conflicts?

**Comment by mautier** [comment]
@jbschlosser Rebased on master and (force-)pushed. Let me know if you want me to squash the commits into one too (not sure if you merge or squash-merge PRs)!

**Comment by mautier** [comment]
Rebased again to pick up 2 fixes that went into master since the last rebase; this should address the `quick-checks` and `shellcheck` lint failures.
